File: leetecode/MaxPower.py
from itertools import accumulate
class FenwickTree:
    def __init__(self, nums):
        self.tree = [0]*(len(nums)+1)

    def update(self, i, diff):
        i += 1
        while i < len(self.tree):
            self.tree[i] += diff
            i += i & (-i)

    def _rangeSum(self, i):
        i += 1
        sum = 0
        while i > 0:
            sum += self.tree[i]
            i -= i & (-i)
        return sum

    def rangeSum(self, i, j):
        return self._rangeSum(j) - self._rangeSum(i-1)
class Solution:
    def maxPower(self, nums: List[int], r: int, k: int) -> int:
        def ok(x):
            temp = k
            acc = FenwickTree(nums)
            for i in range(len(nums)):
                acc.update(i, nums[i])
            for i in range(len(nums)):
                dff = acc.rangeSum(max(0,i-r),min(len(nums)-1,i+r))-x
                # print(dff, temp,i, x)
                if(dff+temp<0): 
                    # print("returning ", False)
                    return False
                if(dff<0):
                    temp += dff
                    acc.update(min(len(nums)-1,i+r),-dff)
            # print("returning ",True)
            return True
        high = 10**8
        low = 0
        while(low<high):
            mid = (low+high)//2;
            if(ok(mid)):
                low = mid+1
            else:
                high = mid
        return low-1

File: leetecode/MaxPower.java
import java.util.*;
public class Solution {
    static class FenwickTree {
        long[] tree;
        public FenwickTree(int[] nums) {
            tree = new long[nums.length+1];
        }
        public void update(int i, long diff) {
            i += 1;
            while (i < tree.length) {
                tree[i] += diff;
                i += i & (-i);
            }
        }
        public long _rangeSum(int i) {
            i += 1;
            long sum = 0;
            while (i > 0) {
                sum += tree[i];
                i -= i & (-i);
            }
            return sum;
        }
        public long rangeSum(int i, int j) {
            return _rangeSum(j) - _rangeSum(i-1);
        }
    }
    public long maxPower(int[] nums, int r, int k) {
        long low = 0;
        long high = (long)Math.pow(10,10);
        while (low < high) {
            long mid = (low+high)/2;
            if (ok(mid, nums, r, k)) low = mid+1;
            else high = mid;
        }
        return low-1;
    }
    public boolean ok(long x, int[] nums, int r, int k) {
        long temp = k;
        FenwickTree acc = new FenwickTree(nums);
        for (int i = 0; i < nums.length; i++) {
            acc.update(i, nums[i]);
        }
        for (int i = 0; i < nums.length; i++) {
            long dff = (long)acc.rangeSum(Math.max(0,i-r),Math.min(nums.length-1,i+r))-x;
            if (dff+temp < 0) return false;
            if (dff < 0) {
                temp += dff;
                acc.update(Math.min(nums.length-1,i+r),-dff);
            }
        }
        return true;
    }
}